README.md
[![NPM version](https://badge.fury.io/js/cdk-automated-waf.svg)](https://badge.fury.io/js/cdk-automated-waf)
[![PyPI version](https://badge.fury.io/py/cdk-automated-waf.svg)](https://badge.fury.io/py/cdk-automated-waf)
[![release](https://github.com/kimisme9386/cdk-automated-waf/actions/workflows/release.yml/badge.svg?branch=main)](https://github.com/kimisme9386/cdk-automated-waf/actions/workflows/release.yml)

# aws-waf-solution

This CDK Construct modify from [Cloudfront with Automated WAF](https://github.com/awslabs/aws-cloudfront-extensions/tree/main/templates/aws-cloudfront-waf).

## Feature

Reference [Cloudfront with Automated WAF README](https://github.com/awslabs/aws-cloudfront-extensions/tree/main/templates/aws-cloudfront-waf#cloudfront-with-automated-waf)

## Construct Props

| Properties                     | Default                      | Description                                                                                                                                                                                                                                                                                                                                                                 |
| ------------------------------ | ---------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **waf2Scope**                  | No                           | `CLOUDFRONT` or `REGIONAL`. If use REGIONALREGIONAL, it support ALB only now                                                                                                                                                                                                                                                                                                |
| **albArn**                     | No                           | if waf2Scope is REGIONAL, it can be specified associated resource to ALB.                                                                                                                                                                                                                                                                                                   |
| **resourceNamingPrefix**       | No                           | If the construct need to deploy more than one times, specify the property to prevent AWS resource name conflict. (The property only allow alphanumeric and "\_" symbol because glue database naming is needed)                                                                                                                                                              |
| **enableShieldAdvancedLambda** | false                        | enable or disable AWS Shield Advance (it need [$3000 Monthly Fee](https://aws.amazon.com/shield/pricing/?nc1=h_ls))                                                                                                                                                                                                                                                         |
| **appAccessLogBucketName**     | access-log-bucket-cloudfront | The name for the Amazon S3 bucket where you want to store Cloud Front access logs for your CloudFront distribution. More about bucket name restriction here: http://amzn.to/1p1YlU5.                                                                                                                                                                                        |
| **logLevel**                   | waf-log-bucket-cloudfront    | The name for the Amazon S3 bucket where you want to store WAF access Cloud Front logs. More about bucket name restriction here: http://amzn.to/1p1YlU5.                                                                                                                                                                                                                     |
| **errorThreshold**             | 50                           | The maximum acceptable bad requests per minute per IP.                                                                                                                                                                                                                                                                                                                      |
| **requestThreshold**           | 100                          | The maximum acceptable requests per FIVE-minute period per IP address.                                                                                                                                                                                                                                                                                                      |
| **blockPeriod**                | 240                          | The period (in minutes) to block applicable IP addresses.                                                                                                                                                                                                                                                                                                                   |
| **WAFScope**                   | CLOUDFRONT                   | Specifies whether this is for an AWS CloudFront distribution or for a regional application. A regional application can be an Application Load Balancer (ALB), an Amazon API Gateway REST API, or an AWS AppSync GraphQL API. Valid Values are CLOUDFRONT and REGIONAL. For CLOUDFRONT, you must create your WAFv2 resources in the US East (N. Virginia) Region, us-east-1. |

## Usage

ALB

```ts
const stack = new cdk.Stack(app, 'TestStackAutomatedWaf', { env });

new AutomatedWaf(stack, 'AutomatedWaf', {
  waf2Scope: Waf2ScopeOption.REGIONAL,
  wafNamingPrefix: 'Alb-Api',
  errorThreshold: 50,
  requestThreshold: 500,
  blockPeriod: 120,
  logLevel: LogLevel.DEBUG,
});
```

:warning: If waf2Scope is REGIONAL, it has to enable ALB access logging and specify S3 location after deployment. The S3 location refer to CloudFormation Output which key is similar to "{stack name}AppAccessLogBucketName{hashCode}". See it in [AWS Document](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html#enable-access-logging)

CloudFront

```ts
const stack = new cdk.Stack(app, 'TestStackAutomatedWaf', { env });

new AutomatedWaf(stack, 'AutomatedWaf', {
  waf2Scope: Waf2ScopeOption.CLOUDFRONT,
  wafNamingPrefix: 'CloudFront-Api',
  errorThreshold: 50,
  requestThreshold: 500,
  blockPeriod: 120,
  logLevel: LogLevel.DEBUG,
});
```

Multiple Stacks

```ts
const stackTest1 = new cdk.Stack(app, 'TestStack1AutomatedWaf', { env });

new AutomatedWaf(stackTest1, 'AutomatedWaf', {
  waf2Scope: Waf2ScopeOption.REGIONAL,
  resourceNamingPrefix: 'Alb-Api',
  errorThreshold: 50,
  requestThreshold: 500,
  blockPeriod: 120,
  logLevel: LogLevel.DEBUG,
});

const stackTest2 = new cdk.Stack(app, 'TestStack2AutomatedWaf', { env });

new AutomatedWaf(stackTest2, 'AutomatedWaf', {
  waf2Scope: Waf2ScopeOption.REGIONAL,
  resourceNamingPrefix: 'Alb-Api2',
  errorThreshold: 60,
  requestThreshold: 600,
  blockPeriod: 120,
  logLevel: LogLevel.DEBUG,
});
```

:warning: If the construct need to deploy more than one times, specify the different value of `resourceNamingPrefix` property to prevent AWS resource name conflict

## Troubleshooting

If deployment error, the cloudFormation Error event like this

```
Received response status [FAILED] from custom resource. Message returned: 'HttpFloodLambdaLogParser' (RequestId: b4e08ea2-fe0a-46f8-98aa-6f96d4558579)
```

If any custom resource deploy error like above, delete the stack and redeploy it that will pass.

## Modified items

- CDK structure refactor from CDK Stack to CDK Construct
- Add optional `enableShieldAdvancedLambda` property on construct because AWS Shield advance need [$3000 Monthly Fee](https://aws.amazon.com/shield/pricing/?nc1=h_ls)
- Lambda deployment use docker image on CDK (DockerImageFunction)
- Fix glue table schema for ALB access log


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.2](https://github.com/kimisme9386/cdk-automated-waf/compare/v0.1.1...v0.1.2) (2021-07-19)

### [0.1.1](https://github.com/kimisme9386/cdk-automated-waf/compare/v0.1.0...v0.1.1) (2021-07-19)


### Bug Fixes

* update snapshot test ([9a9388d](https://github.com/kimisme9386/cdk-automated-waf/commit/9a9388d43b4f3cc848e0fc65f38fc610eef2dc7b))

## [0.1.0](https://github.com/kimisme9386/cdk-automated-waf/compare/v0.0.1...v0.1.0) (2021-07-19)


### Bug Fixes

* update dockerfile and copy lib ([5ff69e2](https://github.com/kimisme9386/cdk-automated-waf/commit/5ff69e228f3dbbade47d1d0ab5f45ba71779237e))

## [0.1.0](https://github.com/kimisme9386/cdk-automated-waf/compare/v0.0.1...v0.1.0) (2021-07-19)


### Bug Fixes

* update dockerfile and copy lib ([5ff69e2](https://github.com/kimisme9386/cdk-automated-waf/commit/5ff69e228f3dbbade47d1d0ab5f45ba71779237e))

## [0.1.0](https://github.com/kimisme9386/cdk-automated-waf/compare/v0.0.1...v0.1.0) (2021-07-19)


test/integ.snapshot.test.ts
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { IntegTesting } from '../src/integ.default';

test('integ snapshot validation', () => {
  const integ = new IntegTesting();
  integ.stack.forEach((stack) => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});


.projenrc.js
const {
  AwsCdkConstructLibrary,
  DependenciesUpgradeMechanism,
} = require('projen');

const AUTOMATION_TOKEN = 'PROJEN_GITHUB_TOKEN';

const project = new AwsCdkConstructLibrary({
  author: 'Chris Yang',
  authorUrl: 'https://9incloud.com/',
  cdkVersion: '1.123.0',
  defaultReleaseBranch: 'main',
  name: '@ikala-cloud/aws-waf-solution',
  repositoryUrl: 'https://github.com/iKalaCloud/aws-waf-solution',
  description: 'Cloudfront and ALB with Automated WAF',
  cdkDependencies: [
    '@aws-cdk/core',
    '@aws-cdk/aws-iam',
    '@aws-cdk/aws-lambda',
    '@aws-cdk/aws-s3',
    '@aws-cdk/aws-kinesisfirehose',
    '@aws-cdk/aws-glue',
    '@aws-cdk/aws-athena',
    '@aws-cdk/aws-wafv2',
    '@aws-cdk/aws-cloudwatch',
    '@aws-cdk/aws-events',
    '@aws-cdk/aws-events-targets',
    '@aws-cdk/aws-s3-notifications',
    '@aws-cdk/custom-resources',
    '@aws-cdk/aws-apigateway',
  ],
  devDeps: ['projen-automate-it'],
  publishToPypi: {
    distName: 'ikala-cloud.aws-waf-solution',
    module: 'ikala-cloud.aws-waf-solution',
  },
  catalog: {
    announce: true,
  },
  stability: 'experimental',
  autoApproveOptions: {
    secret: 'GITHUB_TOKEN',
    allowedUsernames: ['kimisme9386-bot'],
  },
  depsUpgrade: DependenciesUpgradeMechanism.githubWorkflow({
    workflowOptions: {
      labels: ['auto-approve'],
      secret: AUTOMATION_TOKEN,
    },
    ignoreProjen: false,
  }),
  release: false,
  // update snapshot test by workflow because aws region need undefined
  releaseWorkflowSetupSteps: [
    {
      name: 'run build',
      run: 'yarn build',
    },
    {
      name: 'run test',
      run: 'yarn test',
    },
    {
      name: 'run test update',
      run: 'yarn test:update',
    },
    {
      uses: 'EndBug/add-and-commit@v7',
      with: {
        add: 'test',
        author_name: 'Chris Yang',
        author_email: 'kimisme9386@gmail.com',
        message: 'fix: update test snapshot',
      },
    },
  ],
});

project.eslint.addRules({
  'comma-dangle': [
    'error',
    {
      arrays: 'always-multiline',
      objects: 'always-multiline',
      imports: 'always-multiline',
      exports: 'always-multiline',
      functions: 'never',
    },
  ],
});

const common_exclude = [
  'cdk.out',
  'cdk.context.json',
  'images',
  'yarn-error.log',
  'dependabot.yml',
  'lambda/codepipeline-event/dist',
];

project.npmignore.exclude(...common_exclude);
project.gitignore.exclude(...common_exclude);

const common_include = ['/lambda/codepipeline-event/tsconfig.json'];

project.npmignore.include(...common_include);
project.gitignore.include(...common_include);

project.synth();


src/integ.default.ts
import * as cdk from '@aws-cdk/core';
import { AutomatedWaf, LogLevel, Waf2ScopeOption } from './index';

export class IntegTesting {
  readonly stack: cdk.Stack[];
  constructor() {
    const app = new cdk.App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stackTest1 = new cdk.Stack(app, 'TestStack1AutomatedWaf', { env });

    new AutomatedWaf(stackTest1, 'AutomatedWaf', {
      waf2Scope: Waf2ScopeOption.REGIONAL,
      resourceNamingPrefix: 'Alb_Api',
      errorThreshold: 50,
      requestThreshold: 500,
      blockPeriod: 120,
      logLevel: LogLevel.DEBUG,
    });

    const stackTest2 = new cdk.Stack(app, 'TestStack2AutomatedWaf', { env });

    new AutomatedWaf(stackTest2, 'AutomatedWaf', {
      waf2Scope: Waf2ScopeOption.REGIONAL,
      resourceNamingPrefix: 'Alb_Api2',
      errorThreshold: 60,
      requestThreshold: 600,
      blockPeriod: 120,
      logLevel: LogLevel.DEBUG,
    });

    app.synth();
    this.stack = [stackTest1, stackTest2];
  }
}

new IntegTesting();
